### PR TITLE
Add "from_seed" APIs for public key encryption

### PR DESCRIPTION
**Pull request by lmctv**
Should fulfill #292 requirements.

### CODE REVIEW

File: src/nacl/public.py [hunk 1]
```diff
@@ -72,27 +72,46 @@ class PrivateKey(encoding.Encodable, StringFixer, object):
...
+        # Verify the given seed type and size are correct
+        if not (isinstance(seed, bytes) and len(seed) == cls.SEED_SIZE):
+            raise exc.TypeError(("PrivateKey seed must be a {0} bytes long "
+                                 "binary sequence").format(
+                                                           cls.SEED_SIZE)
+                                )
+        # generate a raw keypair from the given seed
+        raw_pk, raw_sk = nacl.bindings.crypto_box_seed_keypair(seed)
+        # create a PrivateKey instance from the raw keypair
+        prvk = super(cls.__class__, cls).__new__(cls)
```
Comment (reaperhulk) [inline src/nacl/public.py]: Why not just use the constructor? Is it a significant performance hit to do the scalar mult to re-derive the public key again?

Comment (lmctv) [inline src/nacl/public.py]: This is just from a dislike of repeated work... scalarmult is fast, but why do it twice in a row?
If you think readabiliy counts more than that, I'll go back to using the standard constructor.

Comment (reaperhulk) [inline src/nacl/public.py]: I would prefer to use the standard constructor here, yeah.

Comment (lmctv) [inline src/nacl/public.py]: OK, will update in a few minutes

Comment (lmctv) [inline src/nacl/public.py]: Change pushed.

**Comment by lmctv** [comment]
@reaperhulk @maqp Added the missing docstring to `from_seed` method. Now this should be really ready.

**Comment by maqp** [comment]
The docstring looks good! Great work once again!

**Comment by lmctv** [comment]
@reaperhulk I think this is ready.

**Comment by lmctv** [comment]
@reaperhulk  I hope both this and #320 are ready. Could you try and find some time to look at them?

File: tests/test_bindings.py [hunk 2]
```diff
@@ -244,3 +244,28 @@ def test_box_seal_wrong_types():
     with pytest.raises(TypeError):
         c.crypto_box_seal_open(
             None, A_pubkey, A_secretkey)
+
+
+def test_box_seed_keypair_reference():
+
+    from box_secret_from_seed_ref import test_seed, test_pk, test_sk
```
Comment (reaperhulk) [inline tests/test_bindings.py]: I don't think these need to be in a separate file, let's just have them be inline in this file along with the comments.

Comment (reaperhulk) [reply tests/test_bindings.py]: Oh, having said that I see they're used in both test_bindings and test_box so nevermind.

Do we have any other python files containing vectors like this? Typically we've made these types of vectors text files that we load and parse.

Comment (lmctv) [reply tests/test_bindings.py]: I think I've begun preparing this PR before #297 got merged, and I didn't want to add another open-coded version of test.utils.read_crypto_test_vectors, which would have been necessary to add the needed comments about the reference data. Hope this is better now.

Comment (lmctv) [reply tests/test_bindings.py]: OK rebased on master after merge of #316

**Pull request merged**